symengine/polynomial.h

```diff
@@ -95,11 +95,230 @@ RCP<const UnivariateIntPolynomial> sub_poly(const UnivariateIntPolynomial &a, co
+    bool is_minus_one() const;
+    //! \return `true` if integer
+    bool is_integer() const;
+    //! \return `true` if symbol
+    bool is_symbol() const;
+    //! \return `true` if mul
+    bool is_mul() const;
+    //! \return `true` if pow
+    bool is_pow() const;
+
+    virtual vec_basic get_args() const;
+
+    inline RCP<const Symbol> get_var() const {
+        return var_;
+    }
+    inline const map_int_Expr& get_dict() const {
+        return dict_;
+    }
+}; //UnivariatePolynomial
+
+//! Adding two UnivariatePolynomial a and b
+RCP<const UnivariatePolynomial> add_uni_poly(const UnivariatePolynomial &a, const UnivariatePolynomial &b);
+//! Negative of a UnivariatePolynomial
+RCP<const UnivariatePolynomial> neg_uni_poly(const UnivariatePolynomial &a);
+//! Subtracting two UnivariatePolynomial a and b
+RCP<const UnivariatePolynomial> sub_uni_poly(const UnivariatePolynomial &a, const UnivariatePolynomial &b);
+//! Multiplying two UnivariatePolynomial a and b
+RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a, RCP<const UnivariatePolynomial> b);
+
+inline RCP<const UnivariatePolynomial> univariate_polynomial(RCP<const Symbol> i, unsigned int deg, const map_int_Expr&& dict) {
+    return make_rcp<const UnivariatePolynomial>(i, deg, std::move(dict));
+}
+
+class UnivariateExprPolynomial {
+private:
+    RCP<const UnivariatePolynomial> poly_;
+public:
+    //! Construct UnivariateExprPolynomial from UnivariatePolynomial
+#if defined(HAVE_SYMENGINE_IS_CONSTRUCTIBLE)
+    template <typename T, typename = typename std::enable_if<std::is_constructible<RCP<const UnivariatePolynomial>, T &&>::value>::type>
+#else
+    template <typename T>
+#endif
+    UnivariateExprPolynomial(T &&o) : poly_(std::forward<T>(o)) {}
+    UnivariateExprPolynomial() {}
+    ~UnivariateExprPolynomial() SYMENGINE_NOEXCEPT {}
+    UnivariateExprPolynomial(const UnivariateExprPolynomial &) = default;
+    UnivariateExprPolynomial(UnivariateExprPolynomial &&other) SYMENGINE_NOEXCEPT : poly_(std::move(other.poly_)) {}
+    UnivariateExprPolynomial(int i) : poly_(UnivariatePolynomial::create(symbol("x"), {{0, Expression(i)}})) {}
+    UnivariateExprPolynomial(RCP<const UnivariatePolynomial> p) : poly_(std::move(p)) {}
+    UnivariateExprPolynomial(Expression expr) : poly_(UnivariatePolynomial::create(symbol("x"), {{0, expr}})) {}
+    UnivariateExprPolynomial &operator=(const UnivariateExprPolynomial &) = default;
+    UnivariateExprPolynomial &operator=(UnivariateExprPolynomial &&other) SYMENGINE_NOEXCEPT {
+        if (this != &other)
+            this->poly_.ptr() = std::move(other.poly_.ptr());
+        return *this;
+    }
+     
+    friend std::ostream &operator<<(std::ostream &os, const UnivariateExprPolynomial &expr) {
+        os << expr.poly_.ptr()->__str__();
+        return os;
+    }
+    
+    friend UnivariateExprPolynomial operator+(const UnivariateExprPolynomial &a, const UnivariateExprPolynomial &b) {
```

Comment:
Is that really necessary? I think add_uni_poly() should take care of those cases?